pkg/volume/csi/csi_plugin.go

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"errors"
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/golang/glog"
+	api "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+	csipb "k8s.io/kubernetes/pkg/volume/csi/proto/csi"
+)
+
+const (
+	csiName       = "csi"
+	csiPluginName = "kubernetes.io/csi"
+	//csiAddrTemplate = "/var/lib/kubelet/plugins/csi/sockets/%v/kubeletproxy.sock"
+	csiAddrTemplate = "/tmp/kubeletproxy.sock"
+	csiTimeout      = 15 * time.Second
+)
+
+var (
+	// csiVersion supported csi version
+	csiVersion = &csipb.Version{Major: 0, Minor: 1, Patch: 0}
+)
+
+type csiPlugin struct {
+	host volume.VolumeHost
+}
+
+// ProbeVolumePlugins returns implemented plugins
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	p := &csiPlugin{
+		host: nil,
+	}
+	return []volume.VolumePlugin{p}
+}
+
+// volume.VolumePlugin methods
+var _ volume.VolumePlugin = &csiPlugin{}
+
+func (p *csiPlugin) Init(host volume.VolumeHost) error {
+	glog.Info(log("plugin initializing..."))
+	p.host = host
+	return nil
+}
+
+func (p *csiPlugin) GetPluginName() string {
+	return csiPluginName
+}
+
+func (p *csiPlugin) GetVolumeName(spec *volume.Spec) (string, error) {
+	csi, err := getCSISourceFromSpec(spec)
+	if err != nil {
+		return "", err
+	}
+	return csi.VolumeHandle, nil
+}
+
+func (p *csiPlugin) CanSupport(spec *volume.Spec) bool {
+	return spec.PersistentVolume != nil && spec.PersistentVolume.Spec.CSI != nil
```

Comment:
Especially this one must be different on nodes (where you should compare also Driver name) and on controllers (where presence of Spec.CSI is enough).

Reply:
@jsafrane so you are saying `CanSupport` should take into consideration the Driver the spec is configured for? If so, agreed.  I will add TODO here.

Reply:
yes, `CanSupport` in kubelet it should check the driver, on the controller-manager it should not.